packages/playground/data-liberation/src/import/WP_Topological_Sorter.php

```diff
@@ -0,0 +1,461 @@
+
+	/**
+	 * The current session ID.
+	 */
+	protected $current_session = null;
+
+	/**
+	 * The total number of categories.
+	 */
+	protected $total_categories = 0;
+
+	/**
+	 * The total number of posts.
+	 */
+	protected $total_posts = 0;
+
+	/**
+	 * The current item being processed.
+	 */
+	protected $current_item = 0;
+
+	public function __construct( $options = array() ) {
+		if ( array_key_exists( 'session_id', $options ) ) {
+			$this->current_session = $options['session_id'];
+		}
+	}
+
+	/**
+	 * Get the name of the table.
+	 *
+	 * @return string The name of the table.
+	 */
+	public static function get_table_name() {
+		global $wpdb;
+
+		// Default is wp_{TABLE_NAME}
+		return $wpdb->prefix . self::TABLE_NAME;
+	}
+
+	/**
+	 * Run by register_activation_hook.
+	 */
+	public static function activate() {
+		global $wpdb;
+
+		$table_name = self::get_table_name();
+
+		// Create the table if it doesn't exist.
+		// @TODO: remove this custom SQLite declaration after first phase of unit tests is done.
+		if ( self::is_sqlite() ) {
+			$sql = $wpdb->prepare(
+				'CREATE TABLE IF NOT EXISTS %i (
+					id INTEGER PRIMARY KEY AUTOINCREMENT,
+					session_id INTEGER NOT NULL,
+					element_type INTEGER NOT NULL default %d,
+					element_id TEXT NOT NULL,
+					parent_id TEXT DEFAULT NULL,
+					byte_offset INTEGER NOT NULL,
+					sort_order int DEFAULT 1
+				);
+
+				CREATE UNIQUE INDEX IF NOT EXISTS idx_element_id ON %i (element_id);
+				CREATE INDEX IF NOT EXISTS idx_session_id ON %i (session_id);
+				CREATE INDEX IF NOT EXISTS idx_parent_id ON %i (parent_id);
+				CREATE INDEX IF NOT EXISTS idx_byte_offset ON %i (byte_offset);',
+				$table_name,
+				self::ELEMENT_TYPE_POST,
+				$table_name,
+				$table_name,
+				$table_name,
+				$table_name
+			);
+		} else {
+			// See wp_get_db_schema
+			$max_index_length = 191;
+
+			// MySQL, MariaDB.
+			$sql = $wpdb->prepare(
+				'CREATE TABLE IF NOT EXISTS %i (
+					id bigint(20) unsigned NOT NULL AUTO_INCREMENT,
+					session_id bigint(20) unsigned NOT NULL,
+					element_type tinyint(1) NOT NULL default %d,
+					element_id text NOT NULL,
+					parent_id text DEFAULT NULL,
+					byte_offset bigint(20) unsigned NOT NULL,
+					sort_order int DEFAULT 1,
+					PRIMARY KEY  (id),
+					KEY session_id (session_id),
+					KEY element_id (element_id(%d)),
+					KEY parent_id (parent_id(%d)),
+					KEY byte_offset (byte_offset)
+				) ' . $wpdb->get_charset_collate(),
+				self::get_table_name(),
+				self::ELEMENT_TYPE_POST,
+				$max_index_length,
+				$max_index_length
+			);
+		}
+
+		require_once ABSPATH . 'wp-admin/includes/upgrade.php';
+		dbDelta( $sql );
+
+		update_option( self::OPTION_NAME, self::DB_VERSION );
+	}
+
+	public static function is_sqlite() {
+		return defined( 'DB_ENGINE' ) && 'sqlite' === DB_ENGINE;
+	}
+
+	/**
+	 * Run in the 'plugins_loaded' action.
+	 */
+	public static function load() {
+		if ( self::DB_VERSION !== (int) get_site_option( self::OPTION_NAME ) ) {
```

Comment:
I'm confused about the distinction between `load()` and `activate()` – could we either have one or document them a bit more to explain the usage?

Reply:
The activation/deactivation is when a user clicks "activate" and "deactivate" link. The load action is usually used if you need to change schema:

The best practices are here:
1. Installation, activation steps: https://developer.wordpress.org/plugins/plugin-basics/activation-deactivation-hooks/
2. Custom tables: https://learn.wordpress.org/lesson/custom-database-tables/
3. Schema change: https://developer.wordpress.org/plugins/creating-tables-with-plugins/

Names can be misleading. Let me add more docs.